### ISSUE TITLE
[REVIEW]: psychTestR: An R package for designing and conducting behavioural psychological experiments

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2088]
**Submitting author:** @pmcharrison (<a href="http://orcid.org/0000-0002-9851-9462">Peter Harrison</a>)
**Repository:** <a href="https://github.com/pmcharrison/psychTestR" target ="_blank">https://github.com/pmcharrison/psychTestR</a>
**Version:** v2.19.0
**Editor:** @oliviaguest
**Reviewer:** @Cognitive-Modeller, @DivyaSeernani
**Archive:** <a href="https://doi.org/10.5281/zenodo.1415355" target="_blank">10.5281/zenodo.1415355</a>

## Status

[![status](https://joss.theoj.org/papers/f76f2883babc0b99ed4960e9fcb0c330/status.svg)](https://joss.theoj.org/papers/f76f2883babc0b99ed4960e9fcb0c330)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/f76f2883babc0b99ed4960e9fcb0c330"><img src="https://joss.theoj.org/papers/f76f2883babc0b99ed4960e9fcb0c330/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/f76f2883babc0b99ed4960e9fcb0c330/status.svg)](https://joss.theoj.org/papers/f76f2883babc0b99ed4960e9fcb0c330)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/pmcharrison/psychTestR" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@Cognitive-Modeller & @DivyaSeernani, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @oliviaguest know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/pmcharrison/psychTestR">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@pmcharrison) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/pmcharrison/psychTestR">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@pmcharrison) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @Cognitive-Modeller, @DivyaSeernani it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.3758/s13428-018-01193-y is OK
- 10.3758/s13428-014-0458-y is OK
- 10.1111/bmsp.12057 is OK
- 10.1038/s41598-017-03586-z is OK
- 10.1038/s41598-018-30318-8 is OK
- 10.3758/s13428-019-01225-1 is OK
- 10.31234/osf.io/8gvhz is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02088/joss.02088/10.21105.joss.02088.pdf)

**Comment by oliviaguest** [comment]
@Cognitive-Modeller, @DivyaSeernani, @pmcharrison thank you all again for submitting/accepting to review this. Start ideally by reading the checklist at the very top of this page and then looking at the code and documentation. :blush: 

If you need to discuss things that are very specific to the code of this submission feel free to open an issue in the [code's repo itself](https://github.com/pmcharrison/psychTestR) and then link back to it from here.

Thank you!  :sparkles:

**Comment by DivyaSeernani** [comment]
Hi there, I couldn't review this week, have a busy week ahead, and will only be able to get to this the week after. Should ideally have the review done by the end of the month.

**Comment by oliviaguest** [comment]
No problem. I'll set up a reminder for both of us. 😊

**Comment by oliviaguest** [comment]
@whedon remind @DivyaSeernani in 10 days

**Comment by whedon** [comment]
Reminder set for @DivyaSeernani in 10 days

**Comment by whedon** [comment]
:wave: @DivyaSeernani, please update us on how your review is going.

**Comment by oliviaguest** [comment]
@DivyaSeernani is this week looking better? :blush:

**Comment by DivyaSeernani** [comment]
Hia, just finished the review. Wasn't sure about the question on the license file since the .txt seems to contain a lot less information than the .md. Also there are clear community guidelines on the website, not in the readme in the repo. I guess that's ok? Otherwise, looks good and I'm looking forward to using the package for future experiments.

**Comment by oliviaguest** [comment]
@Cognitive-Modeller any chance on an ETA for your review, please? :smile: 

@DivyaSeernani fantastic! Thank you so much! :clap: 

@pmcharrison can you check out what @DivyaSeernani is asking about and address as much as you can, please?

**Comment by oliviaguest** [comment]
@Cognitive-Modeller ☝️

**Comment by pmcharrison** [comment]
Hello @DivyaSeernani, thank you very much for this careful checking! 

Regarding the community guidelines, I agree that it would be good to include information in the readme as well as on the website. I've therefore added the following section to the readme:

> The [psychTestR documentation website](https://pmcharrison.github.io/psychTestR/) answers many common questions about psychTestR usage. Further questions are welcomed via the [GitHub issue tracker](https://github.com/pmcharrison/psychTestR/issues). This issue tracker is also the place to report issues or software problems. We welcome software contributions in the form of [GitHub pull requests](https://github.com/pmcharrison/psychTestR/pulls); if you are thinking of making a contribution, consider first beginning a discussion on the issue tracker so that we can discuss implementation details.

Regarding the license file, yes, this reflects an awkward inconsistency between GitHub and CRAN license requirements (see https://stackoverflow.com/questions/43550479/how-to-satisfy-both-cran-and-github-license-file-naming-requirements and https://github.com/tidyverse/tidyr/issues/262 for discussions). Basically, CRAN doesn't want people to include full license texts, but GitHub does (and it's necessary to include the full license when distributing code). The current solution with the two files reflects a common solution implemented in the usethis package (see `usethis::use_mit_license`).

**Comment by oliviaguest** [comment]
@whedon remind @Cognitive-Modeller in 8 days

**Comment by whedon** [comment]
Reminder set for @Cognitive-Modeller in 8 days

**Comment by oliviaguest** [comment]
@Cognitive-Modeller just setting a reminder for you — Wednesday is better but if not this should provide an appropriate reminder for both of us. 😺

**Comment by DivyaSeernani** [comment]
Thanks for the clarification @pmcharrison. I've updated the last incomplete checkbox in my review.

**Comment by oliviaguest** [comment]
@Cognitive-Modeller @DivyaSeernani Thank you both for your reviews so far. I noticed that you both have left this one unchecked:

>  Performance: If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

Is this an ongoing issue and has @pmcharrison addressed this yet? :relaxed:

**Comment by pmcharrison** [comment]
I guess there is the following in the article, which closest approximates a performance claim:

> The psychTestR server model is designed with scalability in mind, using an efficient single-process model that shares memory between sessions and hence easily scales to hundreds of simultaneous participants.

My best evidence for this claim is that we've run these kinds of highly parallel experiments on school students, and seen that scaling works well. A more quantitative test could be done using something like the [shinyloadtest framework](https://rstudio.github.io/shinyloadtest/) - let me know if you think I need to do something like this.

**Comment by oliviaguest** [comment]
> My best evidence for this claim is that we've run these kinds of highly parallel experiments on school students, and seen that scaling works well. A more quantitative test could be done using something like the shinyloadtest framework - let me know if you think I need to do something like this.

@Cognitive-Modeller @DivyaSeernani I leave this up to you to resolve and discuss with @pmcharrison. 👍 When both reviewers have commented/ticked the box, we can continue to the next step. Thanks all for your contributions so far!

**Comment by pmcharrison** [comment]
Hi @Cognitive-Modeller, @DivyaSeernani - what do you think about this performance claim question? Do you think I need to run and report some scaling tests, or do you think it's fine to rely on personal experience using psychTestR in studies?

**Comment by whedon** [comment]
:wave: @Cognitive-Modeller, please update us on how your review is going.

**Comment by oliviaguest** [comment]
@Cognitive-Modeller see @pmcharrison's comment aboive in relation to:

>  Performance: If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

:point_up: :relaxed:

**Comment by DivyaSeernani** [comment]
Hey @pmcharrison I'm fine with experiments that have actually used the package and reporting these studies; have checked the relevant box.

**Comment by arfon** [comment]
Dear authors and reviewers

We wanted to notify you that in light of the current COVID-19 pandemic, JOSS has decided to suspend submission of new manuscripts and to handle existing manuscripts (such as this one) on a "best efforts basis". We understand that you may need to attend to more pressing issues than completing a review or updating a repository in response to a review. If this is the case, a quick note indicating that you need to put a "pause" on your involvement with a review would be appreciated but is not required.

Thanks in advance for your understanding.

_Arfon Smith, Editor in Chief, on behalf of the JOSS editorial team._

**Comment by arfon** [comment]
:wave: @oliviaguest - it looks like both of the reviews are complete here. Is this ready to accept from your perspective?

**Comment by pmcharrison** [comment]
Hi @oliviaguest, @Cognitive-Modeller, @DivyaSeernani - just to add to the above, we had a live test today with 120 simultaneous participants and performance didn't suffer at all, I'm told. We've also had an online data collection over the last couple of months where we tested c. 300,000 participants (though not all at the same time!), and this seemed to work fine. Any chance we could move forward now?

**Comment by oliviaguest** [comment]
> @Cognitive-Modeller see @pmcharrison's comment aboive in relation to:
> 
> > Performance: If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)
> 
> ☝️ ☺️

@pmcharrison I'm hoping @Cognitive-Modeller can report back. I'll send him an email just in case.

**Comment by oliviaguest** [comment]
@arfon IMHO yes, shall we just assume everything is OK and accept it, or wait on @Cognitive-Modeller?

**Comment by oliviaguest** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02088/joss.02088/10.21105.joss.02088.pdf)

**Comment by oliviaguest** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.3758/s13428-018-01193-y is OK
- 10.3758/s13428-014-0458-y is OK
- 10.1111/bmsp.12057 is OK
- 10.1038/s41598-017-03586-z is OK
- 10.1038/s41598-018-30318-8 is OK
- 10.3758/s13428-019-01225-1 is OK
- 10.31234/osf.io/8gvhz is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by oliviaguest** [comment]
@whedon check repository

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.84  T=0.11 s (870.8 files/s, 89520.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
R                               62            461           1577           3705
XML                              1              0             12            686
Markdown                         5             98              0            322
YAML                             2              2              1            163
Rmd                             14            512           2128            138
TeX                              1             10              0             88
JavaScript                      12              3              5             61
-------------------------------------------------------------------------------
SUM:                            97           1086           3723           5163
-------------------------------------------------------------------------------


Statistical information for the repository '2088' was gathered on 2020/05/04.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Peter Harrison                   3            21              1            0.20
Peter M C Harrison              18          5384           5335           99.80

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Peter Harrison               69          328.6         14.6                7.25
```

**Comment by arfon** [comment]
> @arfon IMHO yes, shall we just assume everything is OK and accept it, or wait on @Cognitive-Modeller?

@oliviaguest - yeah, I think we can move ahead with this.

**Comment by oliviaguest** [comment]
@whedon accept

**Comment by whedon** [comment]
No archive DOI set. Exiting...

**Comment by oliviaguest** [comment]
@whedon set v2.19.0 as version

**Comment by whedon** [comment]
OK. v2.19.0 is the version.

**Comment by oliviaguest** [comment]
@pmcharrison can you give me a DOI for your softwar by using, e.g., zenodo, figshare, etc.? And posting the DOI back here, please? 😊

**Comment by pmcharrison** [comment]
Excellent - thanks so much for the swift action! Here's the DOI: 10.5281/zenodo.1415355

**Comment by oliviaguest** [comment]
@whedon set 10.5281/zenodo.1415355 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.1415355" target="_blank">10.5281/zenodo.1415355</a> is the archive.

**Comment by oliviaguest** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.3758/s13428-018-01193-y is OK
- 10.3758/s13428-014-0458-y is OK
- 10.1111/bmsp.12057 is OK
- 10.1038/s41598-017-03586-z is OK
- 10.1038/s41598-018-30318-8 is OK
- 10.3758/s13428-019-01225-1 is OK
- 10.31234/osf.io/8gvhz is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1438

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1438, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by oliviaguest** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02088/joss.02088/10.21105.joss.02088.pdf)

**Comment by oliviaguest** [comment]
@pmcharrison great! I think the final PDF looks really good. What do you think?

**Comment by pmcharrison** [comment]
Thanks! Yes I'm happy :-)

**Comment by oliviaguest** [comment]
@openjournals/joss-eics can somebody please accept this for us? :smile_cat:

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1257461393453965313) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1439
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02088](https://doi.org/10.21105/joss.02088)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by pmcharrison** [comment]
Great! Getting a 404 error on the paper though.

<img width="1184" alt="Screenshot 2020-05-05 at 09 49 51" src="https://user-images.githubusercontent.com/22684998/81049262-c5fae380-8eb5-11ea-9db3-c4bbb5da0e4b.png">

**Comment by oliviaguest** [comment]
@openjournals/joss-eics please see above issue. I get 404 too. 🤔

**Comment by labarba** [comment]
The ping for technical issues should go to @openjournals/dev — Sometimes there are delays on GitHub pages updating. Try again later.

**Comment by oliviaguest** [comment]
@labarba thank you — had no idea! 🌼

**Comment by arfon** [comment]
> @openjournals/joss-eics please see above issue. I get 404 too. 🤔

This looks to be fixed now (you might have to reset your browser cache).

@Cognitive-Modeller, @DivyaSeernani - many thanks for your reviews here and to @oliviaguest for editing this one.

@pmcharrison - your paper is now accepted into JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02088/status.svg)](https://doi.org/10.21105/joss.02088)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02088">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02088/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02088/status.svg
   :target: https://doi.org/10.21105/joss.02088
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02088/status.svg)](https://doi.org/10.21105/joss.02088)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by pmcharrison** [comment]
Thanks very much for all of your work! Very cool workflow.
